the-metalworks/cosmere-rpg
# Feat/flat modifiers to attacks 135

**Type**  

**Description**  
This adds a new input box to the item activation form (for skill roll based activations) that allows users to input a formula and that will get added to the attack roll calculations. Using the existing roll data objects, there are a range of `@` tags for various modifiers and flat numbers and dice can be included to allow for manually representing global effects.

**Related Issue**  
`Closes #135`

**How Has This Been Tested?**  
Loaded build into my local server and tested using GM and Player accounts.
1. Loaded Bridge 9 module into server
2. For the global item modifiers:
    1. Opened the sheet of any of the pre-gens, switch to the inventory tab
    2. Edited a weapon to include a value in the new field
    3. Rolled the item
        - Confirmed the values show up in the formula display
        - Checked that any dice added appear in DSN animation
        - Confirmed that dice values and flat numbers are shown in the chat card and included in the totalling of the outputted value
    4. repeat ii and iii for various scenarios (dice included, use `@mod`/`@attribute`/`@skill.rank`, use `@attr`/`@skills.<skill>.ranks` for any value on the character)
3. For the situational bonuses:
    1. Open the sheet of a pre-gen
    2. Click on an activation icon for any interaction needing a d20 roll
    3. Enter a value in the Temporary Modifier field
        - Confirmed the formula display updated with the new value
    4. Made the Roll
        - Confirmed the roll still works
        - Confirmed DSN still works
        - Confirmed the values showed up in the chat card
        - Confirmed advantage/disadvantage not broken
        - Confirmed shortcut keys not broken
        - Confirmed Attacks don't double-dip the temporary mods
    5. Repeat for various values (+ve and -ve additions, dice, flat numerics, `@` notations, empty field) and for rolls from items as well as weapons and straight skill rolls

**Screenshots (if applicable)**  
Global Mod:

||
|:-:|
|![image](https://github.com/user-attachments/assets/53f468f2-de84-4e52-9232-e4ce6e877883)|
|(I changed the label text since the screenshot, it is now "Roll Formula Extension")<br />⬇️|
|![image](https://github.com/user-attachments/assets/d4ac2536-9cdc-440f-a96f-eed94b21bd76)|
|⬇️|
|![image](https://github.com/user-attachments/assets/4f703eca-c07a-4acc-8f37-ce4c3ee54f91)|
|⬇️|
|![image](https://github.com/user-attachments/assets/d484e2ed-c677-44af-a339-810bb1342e13)|

Temp Mods:

||
|:-:|
|![image](https://github.com/user-attachments/assets/ce077899-e254-4e40-b0a0-9fe0cfc06a47)|
|⬇️|
|![image](https://github.com/user-attachments/assets/85cdf6d9-0390-4b58-9255-dfd0fcec9533)|
|⬇️|
|![image](https://github.com/user-attachments/assets/285847ef-7c2d-4008-9418-f11f7f812449)|


**Checklist:**  
- [x] I have commented on my code, particularly in hard-to-understand areas. *Happy to add more*
